File: backend/src/domain/entities/Emergency.ts
import { v4 as uuidv4 } from 'uuid';

export enum EmergencyType {
  WILDFIRE = 'WILDFIRE',
  EARTHQUAKE = 'EARTHQUAKE',
  FLOOD = 'FLOOD',
  STORM = 'STORM',
  TSUNAMI = 'TSUNAMI',
  LANDSLIDE = 'LANDSLIDE',
  HEATWAVE = 'HEATWAVE',
  OTHER = 'OTHER',
}

export enum EmergencySeverity {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL',
}

export enum EmergencyStatus {
  ACTIVE = 'ACTIVE',
  MONITORING = 'MONITORING',
  RESOLVED = 'RESOLVED',
  CANCELLED = 'CANCELLED',
}

export interface EmergencyProps {
  id?: string;
  communityId: string;
  type: EmergencyType;
  severity: EmergencySeverity;
  status?: EmergencyStatus;
  title: string;
  description: string;
  latitude: number;
  longitude: number;
  radiusKm: number; // affected area
  activatedBy: string; // userId
  resolvedBy?: string;
  createdAt?: Date;
  updatedAt?: Date;
  resolvedAt?: Date;
}

export class Emergency {
  public readonly id: string;
  public communityId: string;
  public type: EmergencyType;
  public severity: EmergencySeverity;
  public status: EmergencyStatus;
  public title: string;
  public description: string;
  public latitude: number;
  public longitude: number;
  public radiusKm: number;
  public activatedBy: string;
  public resolvedBy?: string;
  public readonly createdAt: Date;
  public updatedAt: Date;
  public resolvedAt?: Date;

  constructor(props: EmergencyProps) {
    this.id = props.id ?? uuidv4();
    this.communityId = props.communityId;
    this.type = props.type;
    this.severity = props.severity;
    this.status = props.status ?? EmergencyStatus.ACTIVE;
    this.title = props.title;
    this.description = props.description;
    this.latitude = props.latitude;
    this.longitude = props.longitude;
    this.radiusKm = props.radiusKm;
    this.activatedBy = props.activatedBy;
    this.resolvedBy = props.resolvedBy;
    this.createdAt = props.createdAt ?? new Date();
    this.updatedAt = props.updatedAt ?? new Date();
    this.resolvedAt = props.resolvedAt;
  }

  public isActive(): boolean {
    return this.status === EmergencyStatus.ACTIVE;
  }

  public isCritical(): boolean {
    return this.severity === EmergencySeverity.CRITICAL;
  }

  public escalate(): void {
    const levels = [
      EmergencySeverity.LOW,
      EmergencySeverity.MEDIUM,
      EmergencySeverity.HIGH,
      EmergencySeverity.CRITICAL,
    ];
    const currentIndex = levels.indexOf(this.severity);
    if (currentIndex < levels.length - 1) {
      this.severity = levels[currentIndex + 1];
      this.updatedAt = new Date();
    }
  }

  public resolve(resolvedBy: string): void {
    this.status = EmergencyStatus.RESOLVED;
    this.resolvedBy = resolvedBy;
    this.resolvedAt = new Date();
    this.updatedAt = new Date();
  }

  public cancel(cancelledBy: string): void {
    this.status = EmergencyStatus.CANCELLED;
    this.resolvedBy = cancelledBy;
    this.resolvedAt = new Date();
    this.updatedAt = new Date();
  }

  public expandRadius(additionalKm: number): void {
    this.radiusKm += additionalKm;
    this.updatedAt = new Date();
  }

  public isLocationAffected(lat: number, lng: number): boolean {
    const R = 6371;
    const dLat = ((lat - this.latitude) * Math.PI) / 180;
    const dLng = ((lng - this.longitude) * Math.PI) / 180;
    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos((this.latitude * Math.PI) / 180) *
        Math.cos((lat * Math.PI) / 180) *
        Math.sin(dLng / 2) *
        Math.sin(dLng / 2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c <= this.radiusKm;
  }
}

File: backend-dotnet/src/Hestia.Domain/Entities/Emergency.cs
using Hestia.Domain.Enums;

namespace Hestia.Domain.Entities;

/// <summary>Represents an active emergency event within a community.</summary>
public class Emergency
{
    public Guid Id { get; private set; }
    public Guid CommunityId { get; set; }
    public EmergencyType Type { get; set; }
    public EmergencySeverity Severity { get; set; }
    public EmergencyStatus Status { get; set; }
    public string Title { get; set; } = null!;
    public string Description { get; set; } = null!;
    public double Latitude { get; set; }
    public double Longitude { get; set; }
    public double RadiusKm { get; set; }
    public Guid ActivatedBy { get; set; }
    public Guid? ResolvedBy { get; set; }
    public DateTime? ResolvedAt { get; set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime UpdatedAt { get; set; }

    // Navigation
    public Community? Community { get; set; }
    public ICollection<HelpRequest> HelpRequests { get; set; } = [];
    public ICollection<HelpOffer> HelpOffers { get; set; } = [];

    private Emergency() { }

    public Emergency(
        Guid communityId,
        EmergencyType type,
        EmergencySeverity severity,
        string title,
        string description,
        double latitude,
        double longitude,
        double radiusKm,
        Guid activatedBy)
    {
        Id = Guid.NewGuid();
        CommunityId = communityId;
        Type = type;
        Severity = severity;
        Status = EmergencyStatus.Active;
        Title = title;
        Description = description;
        Latitude = latitude;
        Longitude = longitude;
        RadiusKm = radiusKm;
        ActivatedBy = activatedBy;
        CreatedAt = DateTime.UtcNow;
        UpdatedAt = DateTime.UtcNow;
    }

    public bool IsActive() => Status == EmergencyStatus.Active;

    public bool IsCritical() => Severity == EmergencySeverity.Critical;

    public void Escalate()
    {
        Severity = Severity switch
        {
            EmergencySeverity.Low => EmergencySeverity.Medium,
            EmergencySeverity.Medium => EmergencySeverity.High,
            EmergencySeverity.High => EmergencySeverity.Critical,
            _ => Severity
        };
        UpdatedAt = DateTime.UtcNow;
    }

    public void Resolve(Guid resolvedBy)
    {
        Status = EmergencyStatus.Resolved;
        ResolvedBy = resolvedBy;
        ResolvedAt = DateTime.UtcNow;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Cancel(Guid cancelledBy)
    {
        Status = EmergencyStatus.Cancelled;
        ResolvedBy = cancelledBy;
        ResolvedAt = DateTime.UtcNow;
        UpdatedAt = DateTime.UtcNow;
    }

    public void ExpandRadius(double additionalKm)
    {
        RadiusKm += additionalKm;
        UpdatedAt = DateTime.UtcNow;
    }

    public bool IsLocationAffected(double lat, double lng)
    {
        const double R = 6371;
        var dLat = (lat - Latitude) * Math.PI / 180;
        var dLng = (lng - Longitude) * Math.PI / 180;
        var a = Math.Sin(dLat / 2) * Math.Sin(dLat / 2)
              + Math.Cos(Latitude * Math.PI / 180)
              * Math.Cos(lat * Math.PI / 180)
              * Math.Sin(dLng / 2) * Math.Sin(dLng / 2);
        var c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
        return R * c <= RadiusKm;
    }
}
